qutebrowser-consult.el

```diff
@@ -0,0 +1,222 @@
+;;
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+;; GNU General Public License for more details.
+;;
+;; You should have received a copy of the GNU General Public License
+;; along with this program; if not, write to the Free Software
+;; Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+
+;;; Commentary:
+
+;; Consult-based completion for Qutebrowser buffers, history,
+;; commands, and bookmarks. The sources provided in this file can be
+;; added as additional sources to 'consult-buffer' or similar. See
+;; 'consult-buffer-sources' and 'consult--multi'.
+
+;;; Change Log:
+
+;;; Code:
+(require 'qutebrowser)
+(require 'consult)
+
+(defgroup qutebrowser-consult nil
+  "Consult completion for Qutebrowser."
+  :group 'qutebrowser
+  :prefix "qutebrowser-consult")
+
+;;;; Helper functions
+(defun qutebrowser-consult--format-entry (entry)
+  "Modify ENTRY for presentation."
+  (qutebrowser--shorten-display-url (qutebrowser--tofu-strip entry)))
+
+(defun qutebrowser-consult--annotate (entry)
+  "Return annotation for ENTRY."
+  (qutebrowser--shorten-display-url entry)
+  (propertize (get-text-property 0 'title entry)
+	      'face 'completions-annotations))
+
+;;;; buffer source
+(defvar qutebrowser-consult--exwm-buffer-source
+  (list :name "Qutebrowser buffers"
+        :hidden nil
+        :narrow ?q
+        :history nil
+        :category 'url
+        :action (lambda (entry)
+		  (switch-to-buffer (qutebrowser--tofu-get-buffer entry)))
+        :annotate #'qutebrowser-consult--annotate
+        :items #'qutebrowser-exwm-buffer-search)
+  "Consult source for open Qutebrowser windows.")
+
+;;;; Bookmark source
+(defvar qutebrowser-consult--bookmark-source
+  (list :name "Qutebrowser bookmarks"
+        :hidden nil
+        :narrow ?m
+        :history nil
+        :category 'bookmark
+        :face 'consult-bookmark
+        :action #'qutebrowser-bookmark-jump
+        :items #'qutebrowser-bookmarks-list)
+  "Consult source for Qutebrowser bookmarks.")
+
+;;;; Command source
+(defvar qutebrowser-consult--command-history nil)
+
+(defvar qutebrowser-consult--command-source
+  (list :name "Qutebrowser commands"
+	:hidden nil
+	:narrow ?:
+	:history nil
+	:category 'other
+	:action #'qutebrowser-send-commands
+        :annotate #'qutebrowser-consult--annotate
+	:items (lambda () (mapcar #'qutebrowser--tofu-strip
+				  (qutebrowser-command-search '(":"))))
+	"Consult source for Qutebrowser commands."))
+
+;;;###autoload
+(defun qutebrowser-consult-command (&optional initial)
+  "Command entry for Qutebrowser based on Consult.
+Set initial completion input to INITIAL."
+  (interactive)
+  (let* ((consult-async-min-input 0)
+	 (consult-async-split-style nil)
+	 (selected
+	  (consult--multi `(,(plist-put (seq-copy qutebrowser-consult--command-source)
+					:name nil))
```

Comment:
```suggestion
	  (consult--multi '(qutebrowser-consult--command-source)
                          :group nil
```
This can be simplified by overriding the `:group` function in `consult--multi`.